opts/mod.ts

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2025 Synadia Communications, Inc
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { ConnectionOptions } from "@nats-io/nats-core";
+
+type ConnectionProperty = Omit<
+  keyof ConnectionOptions & { creds: string },
+  "authenticator" | "reconnectDelayHandler"
+>;
+
+const booleanProps = [
+  "debug",
+  "ignoreAuthErrorAbort",
+  "ignoreClusterUpdates",
+  "noAsyncTraces",
+  "noEcho",
+  "noRandomize",
+  "pedantic",
+  "reconnect",
+  "resolve",
+  "verbose",
+  "waitOnFirstConnect",
+  "handshakeFirst",
+] as const;
+
+const numberProps = [
+  "maxPingOut",
+  "maxReconnectAttempts",
+  "pingInterval",
+  "reconnectJitter",
+  "reconnectJitterTLS",
+  "reconnectTimeWait",
+  "timeout",
+] as const;
+
+const stringProps = [
+  "name",
+  "inboxPrefix",
+] as const;
+const arrayProps = ["servers"] as const;
+
+const tlsBooleanProps = [
+  "handshakeFirst",
+];
+
+const tlsStringProps = [
+  "cert",
+  "certFile",
+  "ca",
+  "caFile",
+  "key",
+  "keyFile",
+];
+
+function encodeStringPropsFn(src: Record<string, unknown>, target: URL) {
+  return (n: string) => {
+    const v = src[n] as string | undefined;
+    if (v) {
+      target.searchParams.append(n, encodeURIComponent(v));
+    }
+  };
+}
+
+function encodeBooleanPropsFn(src: Record<string, unknown>, target: URL) {
+  return (n: string) => {
+    const v = src[n] as boolean | undefined;
+    if (typeof v === "boolean") {
+      target.searchParams.append(n, v ? "true" : "false");
+    }
+  };
+}
+
+function encodeNumberPropsFn(src: Record<string, unknown>, target: URL) {
+  return (n: string) => {
+    const v = src[n] as number | undefined;
+    if (typeof v === "number") {
+      target.searchParams.append(n, v.toString());
+    }
+  };
+}
+
+export function encode(opts: Partial<ConnectionOptions>): string {
+  if (typeof opts?.servers === "string") {
+    opts.servers = [opts.servers];
+  }
+  let u: URL;
+  if (opts?.servers?.length) {
+    if (
+      opts.servers[0].startsWith("wss://") ||
+      opts.servers[0].startsWith("ws://")
+    ) {
+      u = new URL(opts.servers[0]);
+    } else {
+      u = new URL(`nats://${opts.servers[0]}`);
```

Comment:
Should this do something to start with the opts.servers starting with something other than `wss://` or `ws://`?  It looks like if I provide `nats://` then this will turn it into `nats://nats://whatever`, and `tls://x` into `nats://tls://x`

Reply:
for the javascript client you don't provide the scheme - it is host port, the only one that needs to have a scheme is `ws/s` or I cannot tell what they are. - But possibly the thing to do is whip through all the servers and rip out the scheme if it isn't ws/s.

Reply:
I have added a check for this - the javascript clients strops `nats/tls` type schemes if they are included as they are meaningless. For secure options, `tls` needs to be set, this prevents inconsistencies where a client connects via TLS on one server, but not in another because of some prefix.

Reply:
So you've now introduced in parse() handling for `natss:` to do the TLS-handshake-on-connect behavior, which is great.  But doesn't this mean that `encode()` should also check for `natss:` and avoid prepending `nats:` to that?  Or are we relying upon the parse inserting only the hostname into the server list, so encode will never see that?

Reply:
Encode is using the connection options, there's no such prefix that is understood - the serialized version does do that, but when decoded recreates the connect options.